enterprise/admin/forms.py

```diff
@@ -5,52 +5,82 @@
 
 from django import forms
-from django.contrib.auth.models import User
 from django.core.exceptions import ValidationError
-from django.core.validators import validate_email
 from django.utils.translation import ugettext as _
 
 from enterprise import utils
-from enterprise.models import EnterpriseCustomer, EnterpriseCustomerUser
+from enterprise.admin.utils import ValidationMessages, username_or_email__to__email, validate_email_to_link
+from enterprise.models import EnterpriseCustomer
 
 
 class ManageLearnersForm(forms.Form):
     """
     Form to manage learner additions.
     """
-    email = forms.CharField(label=_("Email or Username"))
+    username_or_email = forms.CharField(label=_("Email or Username"), required=False)
+    bulk_upload_csv = forms.FileField(label=_("CSV bulk upload"), required=False)
 
-    def clean_email(self):
+    class Modes(object):
+        """
+        Namespace class for form modes.
+        """
+        MODE_SINGULAR = "singular"
+        MODE_BULK = "bulk"
+
+    class Fields(object):
+        """
+        Namespace class for field names.
+        """
+        USERNAME_OR_EMAIL = "username_or_email"
+        BULK_UPLOAD = "bulk_upload_csv"
+        MODE = "mode"
+
+    def clean_username_or_email(self):
         """
         Clean email form field
 
         Returns:
             str: email to link
         """
-        email_or_username = self.cleaned_data["email"].strip()
-        try:
-            user = User.objects.get(username=email_or_username)
-            email = user.email
-        except User.DoesNotExist:
-            email = email_or_username
-
-        try:
-            validate_email(email)
-        except ValidationError:
-            message = _("{email_or_username} does not appear to be a valid email or known username").format(
-                email_or_username=email_or_username
-            )
-            raise ValidationError(message)
-
-        existing_record = EnterpriseCustomerUser.objects.get_link_by_email(email)
-        if existing_record:
-            message = _("User with email {email} is already registered with Enterprise Customer {ec_name}").format(
-                email=email, ec_name=existing_record.enterprise_customer.name
-            )
-            raise ValidationError(message)
+        email_or_username = self.cleaned_data[self.Fields.USERNAME_OR_EMAIL].strip()
+
+        if not email_or_username:
+            return email_or_username
+
+        email = username_or_email__to__email(email_or_username)
+        validate_email_to_link(email, email_or_username)
 
         return email
 
+    def clean(self):
+        """
+        Clean fields that depend on each other.
+
+        In this case, the form can be used to link single user or bulk link multiple users. These are mutually
+        exclusive modes, so this method checks that only one field is passed.
+        """
+        cleaned_data = super(ManageLearnersForm, self).clean()
```

Comment:
I think it's first since it calls the superclass method, and generally calls to `super()` go first. However, in this case, the superclass method is just `def clean(self): return self.cleaned_data` so it doesn't really matter.